### PR TITLE
chore(DSTEW-1580): Add in validation package

### PR DESCRIPTION
**Pull request by BenJusticeDigital**
## Summary

[Link to story](https://dsdmoj.atlassian.net/browse/DSTEW-1580)

### Summary
Introduced a new validation component within the validation package of the Event Service, including a comparison mechanism for both claims and submissions. This comparator evaluates the output of the new validation logic against the existing implementation.

The comparison is executed within a try-catch block to ensure that any failures do not impact current behaviour. Any discrepancies are logged for manual review, allowing us to safely assess differences without affecting production logic.
Additionally, a suite of integration tests has been added to cover a wide range of claim and submission scenarios. These tests are designed to exercise various combinations and edge cases, enabling thorough validation and comparison of results.

### Next Steps
Once confidence in the new validation logic is established, the comparison layer can be removed while retaining the validated checks. At that point, we will be in a position to switch fully to the new validation approach.
## Checklist

Before you ask people to review this PR, please confirm:

- [x] The build pipeline is passing.
- [x] There are no conflicts with the target branch.
- [x] There are no whitespace-only changes.
- [x] There are no unexpected changes.

### CODE REVIEW

File: data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java [hunk 1]
```diff
@@ -50,14 +59,16 @@ public class ClaimValidationService {
...
+   * @param validationService the new validation service
+   * @param categoryOfLawValidationService the category of law validation service
+   * @param dataClaimsRestClient the data claims rest client
+   * @param eventServiceMetricService the event service metric service
+   * @param bulkClaimUpdater the bulk claim updater
+   * @param claimValidator the list of claim validators
+   * @param claimValidationBatchSize the batch size of claims to validate at once
    */
   public ClaimValidationService(
+      ValidationService validationService,
```
Comment (vvartakMoJ) [inline data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java:72]: The presence of the categoryOfLawValidationService makes me question the name of this service (and the javadoc for it which says "the new validation service"). If the other one is still needed, I think we should either rename that or this, or at least the comment, to say what this is (it won't be new for long!)

File: data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java [hunk 2]
```diff
@@ -274,4 +286,64 @@ private void recordClaimMetrics(ClaimResponse claim, SubmissionValidationContext
...
+
+  /**
+   * Maps a ClaimResponse to the internal validation Claim model, runs the new validation service,
+   * and compares the results against the existing report. Differences are logged as WARN via {@link
+   * ValidationResultComparator}. Any unexpected error is caught and logged to prevent disruption to
+   * the existing validation flow.
+   */
+  private void compareClaimValidationResults(
+      ClaimResponse claimResponse,
+      SubmissionResponse submission,
```
Comment (vvartakMoJ) [inline data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java]: Nit: call this submissionResponse too (in line with claimResponse above)

File: data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java [hunk 3]
```diff
@@ -274,4 +286,64 @@ private void recordClaimMetrics(ClaimResponse claim, SubmissionValidationContext
...
+      SubmissionValidationContext context,
+      List<ClaimResponse> finalSubmissionClaims) {
```
Comment (vvartakMoJ) [inline data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java]: Nit: the calling code (earlier in the same class) removed the "final" prefix from this object's name, perhaps remove it from here too)?

File: data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java [hunk 4]
```diff
@@ -118,17 +130,17 @@ public void validateAndUpdateClaims(
           categoryOfLawValidationService.getFeeDetailsResponseForAllFeeCodesInClaims(
               submissionClaims);
 
-      List<ClaimResponse> finalSubmissionClaims = submissionClaims;
-
       // Submit validation tasks for each claim
       for (ClaimResponse claim : submissionClaims) {
         validateClaim(
```
Comment (vvartakMoJ) [inline data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java:136]: Refactor suggestion for your consideration: currently we are passing each claim and the whole list (submissionClaims) which also contains that claim, to validateClaim in a loop.

How about moving the for loop inside that method (and call it validateClaims, i.e. plural), like this:
```
private void validateClaims(
    List<ClaimResponse> submissionClaims,
    Map<String, FeeDetailsResponseWrapper> feeDetailsResponseMap,
    SubmissionResponse submission,
    SubmissionValidationContext context)
```

I think it will read better. The call to compareClaimValidationResults will also have to move to inside that method, but I think that's OK as it is only temporary.

File: data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/util/ValidationResultComparator.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    removeExactMatches(unmatched, unmatchedExisting);
+
+    if (unmatched.isEmpty() && unmatchedExisting.isEmpty()) {
+      log.debug("{} validators matched exactly", label);
+      return;
+    }
+
+    unmatched.forEach(
+        ni -> log.warn("[VALIDATOR-DRY_RUN] {} only in new validator: {}", label, describeNew(ni)));
```
Comment (vvartakMoJ) [inline data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/util/ValidationResultComparator.java]: Typo: should be VALIDATOR-DRY-RUN (hyphen not underscore), to match the other message.

File: data-claims-event-service/src/main/resources/application.yml [hunk 6]
```diff
@@ -20,13 +20,9 @@ logging:
...
   structured:
     format:
-      console: ecs
-    ecs:
-      service:
-        name: laa-data-claims-event-service
-        version: ${spring.application.version:default}
-        environment: ${spring.profiles.active:default}
-        node_name: ${HOSTNAME:unknown}
+      console: "" # disables structured logging locally
```
Comment (vvartakMoJ) [inline data-claims-event-service/src/main/resources/application.yml]: Can you please confirm and mention here if/how the console: ecs get applied in higher environments? I don't see any relevant overrides in helm values). Also, please check the comments in README.md to ensure they are still up to date.

**Review by vvartakMoJ — CHANGES_REQUESTED** [review]
Looks good overall, left some review comments/suggestions.

File: data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java (hunk 1 above)
Comment (BenJusticeDigital) [reply data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java:72]: categoryOfLawValidationService is the one that's out of place it is using a sub method to fetch data for efficiency and is not actually doing any validation. Have added a comment to that effect.

File: data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java [hunk 7]
```diff
@@ -274,4 +287,64 @@ private void recordClaimMetrics(ClaimResponse claim, SubmissionValidationContext
...
+      mappedClaim.setOfficeAccountNumber(submissionResponse.getOfficeAccountNumber());
+
+      List<Claim> relatedClaims =
+          submissionClaims.stream().map(ClaimMapper::fromClaimResponse).toList();
+
+      relatedClaims.forEach(
+          c -> {
+            c.setAreaOfLaw(submissionResponse.getAreaOfLaw());
+            c.setOfficeAccountNumber(submissionResponse.getOfficeAccountNumber());
+          });
```
Comment (vvartakMoJ) [inline data-claims-event-service/src/main/java/uk/gov/justice/laa/dstew/payments/claimsevent/service/ClaimValidationService.java]: Performance related question/suggestion:
`relatedClaims` are calculated based on data that does not change for each claim (it comes from the parent submission). They don't have to be calculated for each claim individually and should be calculated only once for the whole set, before the call to this method (`compareClaimValidationResults`). Assuming there are 100 claims in a page, this will reduce the number of times they calculated from 100 to 1 (That's 10K ClaimMapper calls per page reduced to 100).

**Review by vvartakMoJ — CHANGES_REQUESTED** [review]
One performance related question/suggestion, rest looks good.

**Review by vvartakMoJ — APPROVED** [review]
LGTM!
